### PR TITLE
USVM API implementation for TypeScript

### PR DESCRIPTION
**Pull request by zishkaz**
Basic USVM API implementation for futher development

### CODE REVIEW

File: usvm-ts/src/main/kotlin/org/usvm/TSInterpreter.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        return scope.stepResult()
+    }
+
+    fun getInitialState(method: EtsMethod, targets: List<TSTarget>): TSState {
+        val state = TSState(ctx, method, targets = UTargetsSet.from(targets))
+
+        with(ctx) {
+            val params = method.parameters.mapIndexed { idx, _ ->
```
Comment (CaelmBleidd) [inline usvm-ts/src/main/kotlin/org/usvm/TSInterpreter.kt:55]: Can be replaced with initializer

File: usvm-ts/src/main/kotlin/org/usvm/TSMachine.kt [hunk 2]
```diff
@@ -1,11 +1,126 @@
...
+
+        val methodsToTrackCoverage =
+            when (options.coverageZone) {
+                CoverageZone.METHOD,
+                CoverageZone.TRANSITIVE,
+                -> methods.toSet()
+                // TODO: more adequate method filtering. !it.isConstructor is used to exclude default constructor which is often not covered
+                CoverageZone.CLASS -> methods.mapNotNull { method ->
+                    project.getMethodBySignature(method.signature)
+                }.toSet() + methods
```
Comment (CaelmBleidd) [inline usvm-ts/src/main/kotlin/org/usvm/TSMachine.kt:46]: This code doesn't make sense

File: usvm-ts/src/main/kotlin/org/usvm/TSMachine.kt [hunk 3]
```diff
@@ -1,11 +1,126 @@
...
+
+        val pathSelector = createPathSelector(
+            initialStates,
+            options,
+            applicationGraph,
+            timeStatistics,
+            { coverageStatistics },
+            { transparentCfgStatistics },
+            { callGraphStatistics },
+            { loopTracker }
```
Comment (CaelmBleidd) [inline usvm-ts/src/main/kotlin/org/usvm/TSMachine.kt:72]: We are not sure about loop tracker, lets not include so complicated parts in the initial commits

File: usvm-ts/src/main/kotlin/org/usvm/TSMachine.kt [hunk 4]
```diff
@@ -1,11 +1,126 @@
...
+
+        return statesCollector.collectedStates
+    }
+
     override fun close() {
-        TODO("Not yet implemented")
+        components.close()
+    }
+
+    private fun transparentCfgStatistics() = object : CfgStatistics<EtsMethod, EtsStmt> {
```
Comment (CaelmBleidd) [inline usvm-ts/src/main/kotlin/org/usvm/TSMachine.kt:117]: Transparent statistic has no purpose in the ts code rn

File: usvm-ts/src/main/kotlin/org/usvm/state/TSState.kt [hunk 5]
```diff
@@ -3,11 +3,7 @@ package org.usvm.state
 import org.jacodb.panda.dynamic.ets.base.EtsStmt
 import org.jacodb.panda.dynamic.ets.base.EtsType
 import org.jacodb.panda.dynamic.ets.model.EtsMethod
-import org.usvm.PathNode
-import org.usvm.TSContext
-import org.usvm.TSTarget
-import org.usvm.UCallStack
-import org.usvm.UState
+import org.usvm.*
```
Comment (CaelmBleidd) [inline usvm-ts/src/main/kotlin/org/usvm/state/TSState.kt:6]: No asterisk

File: usvm-ts/src/test/kotlin/org/usvm/util/TSTestResolver.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import org.jacodb.panda.dynamic.ets.model.EtsMethod
+import org.usvm.TSTest
+import org.usvm.state.TSState
+
+@Suppress("UNUSED_PARAMETER", "UNUSED_VARIABLE")
+class TSTestResolver {
+
+    fun resolve(method: EtsMethod, state: TSState): TSTest {
+
+        return TSTest(emptyList(), null)
```
Comment (CaelmBleidd) [inline usvm-ts/src/test/kotlin/org/usvm/util/TSTestResolver.kt:12]: Lets keep not implemented things with todos

File: usvm-ts/src/test/resources/ir/MinValue.json [hunk 7]
```diff
@@ -0,0 +1,636 @@
+{
```
Comment (CaelmBleidd) [inline usvm-ts/src/test/resources/ir/MinValue.json:1]: Its ok for now, but I don't want to have these jsons in the repository in the main branch. It should be replaced with jacodb automatic solution

**Review by CaelmBleidd — CHANGES_REQUESTED** [review]

File: usvm-ts/build.gradle.kts [hunk 8]
```diff
@@ -8,6 +8,9 @@ dependencies {
     implementation("${Versions.jacodbPackage}:jacodb-core:${Versions.jacodb}")
     implementation("${Versions.jacodbPackage}:jacodb-panda-dynamic:${Versions.jacodb}")
 
+    // For some reason USVM TS won't run without this explicit kotlinx dependency
+    api("org.jetbrains.kotlinx:kotlinx-collections-immutable:0.3.7")
```
Comment (Lipen) [inline usvm-ts/build.gradle.kts:12]: See https://github.com/UnitTestBot/usvm/pull/190

**Review by CaelmBleidd — APPROVED** [review]

**Pull request merged**